src/gt4py/next/otf/compiled_program.py

```diff
@@ -360,9 +385,15 @@ def _compile_variant(
             kwargs=self.program_type.definition.kw_only_args,
             argument_descriptor_contexts=argument_descriptor_contexts,
         )
-        self._compiled_programs[key] = _async_compilation_pool.submit(
-            self.backend.compile, self.definition_stage, compile_time_args=compile_time_args
-        )
+        if _async_compilation_pool is None:
+            # synchronous compilation
+            self._compiled_programs[key] = self.backend.compile(
+                self.definition_stage, compile_time_args=compile_time_args
+            )
+        else:
+            self._compiled_programs[key] = _async_compilation_pool.submit(
+                self.backend.compile, self.definition_stage, compile_time_args=compile_time_args
+            )
```

Comment:
I would try to avoid duplicating the compile call logic using something like `functools.partial`:
```suggestion
        compile_call = functools.partial(
            self.backend.compile, self.definition_stage, compile_time_args=compile_time_args
        )
        if _async_compilation_pool is None:
            # synchronous compilation
            self._compiled_programs[key] = compile_call()
        else:
            self._compiled_programs[key] = _async_compilation_pool.submit(compile_call)
```